core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -142,7 +144,83 @@ object ReassignPartitionsCommand extends Logging {
         TopicAndPartition(topic, partition) -> replicas
       }
     }
-    (partitionsToBeReassigned, currentAssignment)
+    (partitionsToBeReassigned, currentAssignment, disableRackAware match {
+      case true => None
+      case false => checkBrokerBalance(brokerMetadatas, partitionsToBeReassigned)
+    })
+  }
+
+  def checkBrokerBalance(brokerMetadatas: Seq[BrokerMetadata], assignment: Map[TopicAndPartition, Seq[Int]]): Option[String] = {
+    val brokerCount = brokerMetadatas.size
+
+    // map of broker id -> rack id
+    val brokerToRackMap = brokerMetadatas.collect { case BrokerMetadata(id, Some(rack)) =>
+      id -> rack
+    }.toMap
+
+    // sorted broker ids
+    val brokers = brokerToRackMap.keys.toList.sortWith(_ < _)
+
+    // map of rack id -> array of broker ids
+    val rackToBrokersMap = brokerToRackMap groupBy{_._2} map {
+      case (key, value) => (key, value.unzip._1.toArray)
+    }
+
+    // map of broker id -> number of topic partition replicas in the broker
+    var brokerToReplicaCountMap = assignment.values.toSet.flatten.map(v => (v, assignment.keys.filter(assignment(_).contains(v)).size)).toMap
+    if (brokerToReplicaCountMap.size < brokerCount) {
+      brokers.toSet.diff(brokerToReplicaCountMap.keySet).foreach { b =>
+        brokerToReplicaCountMap = brokerToReplicaCountMap + (b -> 0)}
+    }
+
+    // map of rack id -> number of topic partition replicas in the rack
+    val rackToReplicaCountMap = rackToBrokersMap map {
+      case (rack, brokers) => {
+        (rack, brokerToReplicaCountMap.filterKeys(brokers.contains).foldLeft(0)(_+_._2))
```

Comment:
Thanks for catching this. I noticed two instances. One with the `+` in this line, and another with a `/` a few lines below. If there is any other that I missed, please let me know.